Waku v2: History response pagination specs

staheri14 opened the pull request:
This PR contains the our current pagination technique as well as the protobuf specifications of  the pagination-related data types and messages. https://github.com/vacp2p/specs/issues/195

oskarth commented:
I think it would make sense to have this as part of the store protocol spec. The reason being that you are probably never going to use pagination in isolation without using the store protocol, and you also pretty much need it for interacting with the store protocol.

Practically, it'd mean a new section in store spec and editing protobufs and changelog.

oskarth commented on specs/waku/v2/waku-store-pagination.md:
```diff
@@ -0,0 +1,102 @@
+title: Waku
+version: 2.0.0-beta1
+status: Draft
+authors: 
```
Add your name (but to the store protocol spec)

oskarth reviewed (commented):
Overall looks good! Some minor formatting stuff. I'll look closer once it is a diff vs the original store protocol

oskarth commented on specs/waku/v2/waku-store-pagination.md:
````diff
@@ -0,0 +1,102 @@
+## Protobuf
+
+```protobuf
+message Index {
````
These would be part of the https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-store.md#protobuf section

staheri14 replied:
Sure

oskarth commented on specs/waku/v2/waku-store-pagination.md:
```diff
@@ -0,0 +1,102 @@
+
+# Abstract
+
+Pagination is an optional service enabled on top of the `WakuStore` protocol. By using pagination, a requester can set the number of waku messages to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
```
Could also be phrased as something like:

Nodes supporting store protocol MAY (or perhaps SHOULD) support Pagination. When a node supports pagination, foo bar baz...

staheri14 replied:
Sure, will modify it

oskarth commented on specs/waku/v2/waku-store-pagination.md:
```diff
@@ -0,0 +1,102 @@
+
+message HistoryQuery {
+  repeated string topics = 2;
+  optional PagingInfo pagingInfo = 3
```
I think we should be explicit about protobuf version here. I believe in the existing store spec https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-store.md#protobuf all fields are optional by default?

cc @decanus who knows better

decanus replied:
yea, it should be defined in the store spec as proto3 where all fields are optional.

decanus reviewed (changes requested):
@oskarth I think it would be best when this spec is combined with the store spec, wdyt?

oskarth commented:
> I think it would make sense to have this as part of the store protocol spec.

Yes, agreed

staheri14 commented:
@oskarth @decanus I added the pagination specs as part of the store protocol. I almost did not touch the description of the store protocol, except for the protobuf part. Please let me know if any updates on the store protocol may be needed.

oskarth commented on specs/waku/v2/waku-store-pagination.md:
```diff
@@ -0,0 +1,102 @@
+---
```
Leftover file I think?

staheri14 replied:
right, will be deleted

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
 title: Waku
 version: 2.0.0-beta1
```
Bump to beta2

oskarth replied:
And add changelog

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
 # Table of Contents
 
 - [Abstract](#abstract)
+  * [Waku Store Protocol](#Waku-Store-Protocol)
```
Usually this is autogenerated with markdown-toc and result would be lower-case-links. Easier to avoid mistakes and do less with this tool

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
 
 # Abstract
 
+## Waku Store Protocol 
```
```suggestion
```
I think this still makes sense as an abstract, where pagination is one aspect of the store protocol. Otherwise they sound disjointed to me

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
 `WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes.
 
+## Pagination
+Nodes supporting store protocol MAY support Pagination. When a node supports pagination, a requester node can set the number of waku messages to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
```
```suggestion
## Pagination

Nodes supporting the store protocol MAY support pagination. When a node supports pagination, a requester node can set the number of `WakuMessage`s to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
```

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
+## Pagination
+Nodes supporting store protocol MAY support Pagination. When a node supports pagination, a requester node can set the number of waku messages to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
+
+To perform pagination, each `WakuMessage` stored at a waku store node is associated with a unique `Index`.  Once a `WakuMessage` is received by a node, it will generate the `Index` of the  `WakuMessage` and then saves the (`Index`, `WakuMessage`) together in its local storage. An `Index` encapsulates fields useful for paging, these fields are a `digest` and a `receivedTime`. The `receivedTime` in ...
```
```suggestion
To perform pagination, each `WakuMessage` stored at a Waku node running the store protocol is associated with a unique `Index`. Once a `WakuMessage` is received by a node, it will generate the `Index` of the  `WakuMessage` and then saves the (`Index`, `WakuMessage`) together in its local storage. An `Index` encapsulates fields useful for paging, these fields are a `digest` and a `receivedTime`. The `receivedTime` indicates the time at which the `WakuMessage` is received by the store node and the `digest` is the hash of the received `WakuMessage`.  
```

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
 
+## Waku Store Protocol 
 `WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes.
 
```
```suggestion
`WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes. It also support pagination for more efficient querying of historical messages.

```

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
+## Waku Store Protocol 
 `WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes.
 
+## Pagination
```
I'd probably pull this down to under wire specification or as part of it. So it'd be something like:

Abstract
Motivation (currently missing and probably out of scope for PR)
Wire spec
 - details about fields
   - possibly on two levels if needed

E.g. something like this https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-relay.md#rpc

oskarth replied:
Right now it seems like it is a lot of duplication, I might be wrong and we need a separate section (e.g. before or after wire spec), but I think it can be done as part of it.

oskarth replied:
If we want to add motivation, it'd be something about nodes being offline a lot and wanting to get historical messages, etc. E.g. how to address subset of https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-v2.md#motivation-and-goals

staheri14 replied:
Sure, I'll include the motivation, perhaps it should cover motivation for the store protocol as well as the pagination, right?

staheri14 replied:
In the abstract section, my aim was to provide an overview of the pagination process and to give an insight into 
why the protobuf messages that are provided in the wire spec are even needed. I thought it would be easier to understand the whole spec this way. However, I agree that the current writeup has duplications, and the wire specs or the abstract must be shorten.

oskarth replied:
I see. To me, the abstract should be quite concise and max 200 words (probably less in this case because it isn't a very complex protocol) without too much technical detail. If that can be achieved without detailing implementation details it seems fine to have a sentence or three re pagination in there.

oskarth replied:
Kind of like an executive summary to get a high level overview of what the store protocol is and why one should care

staheri14 replied:
> I'd probably pull this down to under wire specification or as part of it. So it'd be something like:
> 
> Abstract
> Motivation (currently missing and probably out of scope for PR)

I will open a PR for the motivation

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
+The `HistoryResponse` will also contain an updated `pagingInfo` which helps the requester to resume its further requests; the `pageSize` indicates the number of returned waku messages (i.e., the number of messages included in `messages` field of `HistoryResponse`), the `direction` is the same direction as in the corresponding `HistoryQuery`. The `cursor` holds the `Index` of the last/first messag ...
+
+
 **Protocol identifier***: `/vac/waku/store/2.0.0-beta1`
```
Should be in abstract for easy reference I think

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +93,47 @@ message HistoryRPC {
+A node MUST send all History messages (`HistoryQuery`, `HistoryResponse`) wrapped inside a
+`HistoryRPC`.
+
+- The `request_id` MUST be a uniquely generated string, the `HistoryResponse` and `HistoryQuery` `request_id` MUST match. 
```
Formatting seems off, either - for each, newline or as one paragraph

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +93,47 @@ message HistoryRPC {
+### HistoryRPC
+
+A node MUST send all History messages (`HistoryQuery`, `HistoryResponse`) wrapped inside a
+`HistoryRPC`.
```
```suggestion
A node MUST send all History messages (`HistoryQuery`, `HistoryResponse`) wrapped inside a `HistoryRPC`.
```

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +93,47 @@ message HistoryRPC {
-A node MUST send all History messages (`HistoryQuery`, `HistoryResponse`) wrapped inside a
-`HistoryRPC`.
+### PagingInfo
+`PagingInfo` has three components defined below.
```
```suggestion

### PagingInfo
`PagingInfo` has three components defined below.
```

Ditto above

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
+
+- Is a boolean value
+     - True means forward pagination that is the `messages` field of the `HistoryResponse` contains `pageSize` many (or less) waku messages whose `Index` are larger than the given `cursor`
+     - False means backward pagination, that is the `messages` field of the `HistoryResponse` contains  `pageSize` many (or less) waku messages whose `Index` are smaller than the given cursor
```
```suggestion
- Is a boolean value
     - True means forward pagination that is the `messages` field of the `HistoryResponse` contains `pageSize` many (or less) waku messages whose `Index` are larger than the given `cursor`
     - False means backward pagination, that is the `messages` field of the `HistoryResponse` contains  `pageSize` many (or less) waku messages whose `Index` are smaller than the given cursor
```

I think this can  probably be written more concisely in one or two sentences

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
+     - True means forward pagination that is the `messages` field of the `HistoryResponse` contains `pageSize` many (or less) waku messages whose `Index` are larger than the given `cursor`
+     - False means backward pagination, that is the `messages` field of the `HistoryResponse` contains  `pageSize` many (or less) waku messages whose `Index` are smaller than the given cursor
+
+Given a valid  `pagingInfo`, the waku store node performs as follows:
```
This seems like a HOW vs WHAT. I think as long as we specfiy WHAT is returned the HOW will be an implementation detail, or be immediately apparent (e..g sorting, etc)

staheri14 replied:
I am wondering whether the execution details, like this one, will be ever included in any doc? for example, in the .md files of the nim-waku repo? Or explaining these details is outside of the scope (details of the pagination may seem simple, but I am asking for general cases)?

oskarth replied:
I think either in-line as part of code comments, e.g. like https://github.com/status-im/nim-waku/blob/master/waku/protocol/v1/waku_protocol.nim#L11-L38 or if it is more of an API thing, with markdown files in doc folder like https://github.com/status-im/nim-waku/blob/master/docs/api/v2/node.md

That said, docs are currently lacking and open to ideas for how to best improve them!

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,88 @@
+- Extracts `pageSize` many waku messages whose `Index` exceed/precede the `cursor` (depending on the `direction` being true/false)
+- the resultant list  is sent in the  `messages` field of the `HistoryResponse`
+
+The `HistoryResponse` will also contain an updated `pagingInfo` which helps the requester to resume its further requests; the `pageSize` indicates the number of returned waku messages (i.e., the number of messages included in `messages` field of `HistoryResponse`), the `direction` is the same direction as in the corresponding `HistoryQuery`. The `cursor` holds the `Index` of the last/first messag ...
```
For true/false statements, I'd probably write it explicitly with one case, and then write something like "...and vice versa for direction being false" or "...and true and false respectively for foo".

oskarth replied:
Might want to add a sentence or paragraph on how this allows a querying node to receive all messages in an iterated manner, either starting from the beginning or from the end.

Also, probably belongs further up as a general comment, all of this assumes the node was online to receive messages from relay protocol. There could still be gaps :<

staheri14 replied:
By an online node you meant the responder node is supposed to be online and receive all the messages, right?Then sure, gaps are inevitable, indeed we never attempted this problem (gaps) in the current pagination method, the current design only works for one requester and responder. Please let me know if a separate issue is needed to address the offline (store) nodes?

oskarth replied:
Yeah, I think just a sentence or so to make it clear to the reader what the assumption is, since it might not be obvious to everyone.

I think there might already be an issue for the general store high availability requirement problem, (@decanus might know), but feel free to create one if you want! Definitely not a part of the scope of the current work though, as you say.

oskarth commented:
@decanus I think you forgot to review this one too?

oskarth commented:
@staheri14 let me know when you want me to do a re-review, and/or convert from Draft to Ready for review

decanus commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,94 @@
+
+## Pagination
+
+Nodes supporting the store protocol MAY support pagination. When a node supports pagination, a requester node can set the number of `WakuMessage`s to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
```
```suggestion
Nodes supporting the store protocol MAY support pagination. When a node supports pagination, a querying node can set the number of `WakuMessage`s to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
```

decanus commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,94 @@
+
+Nodes supporting the store protocol MAY support pagination. When a node supports pagination, a requester node can set the number of `WakuMessage`s to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
+
+To perform pagination, each `WakuMessage` stored at a Waku node running the store protocol is associated with a unique `Index`. Once a `WakuMessage` is received by a node, it will generate the `Index` of the  `WakuMessage` and then saves the (`Index`, `WakuMessage`) together in its local storage. An `Index` encapsulates fields useful for paging, these fields are a `digest` and a `receivedTime`. T ...
```
So I think that the `Index` is an implementation detail and therefore not required to be in the spec, wdyt @oskarth?

oskarth replied:
Index is a protobuf field so it should be explained to same extent (type, basic semantics). Agree it can be more minimal though (see duplication comment)

decanus commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +99,46 @@ message HistoryRPC {
 
 2.0.0-beta1
-Initial draft version. Released [2020-10-06](https://github.com/vacp2p/specs/commit/75b4c39e7945eb71ad3f9a0a62b99cff5dac42cf)
+Initial draft version. Released [2020-10-21](https://github.com/vacp2p/specs/commit/75b4c39e7945eb71ad3f9a0a62b99cff5dac42cf)
```
New version

staheri14 replied:
addressed

decanus replied:
@staheri14 there should still be the old version, don't remove the old one. Just add a new one

decanus commented on specs/waku/v2/waku-store.md:
```diff
@@ -2,36 +2,94 @@
+
+**Direction**
+
+- Is a boolean value
```
@staheri14 to limit the need of all this explanation, it probably makes it easier to just make direction an enum

staheri14 replied:
done

staheri14 commented:
@oskarth @decanus I reorganized and shortened the spec, removed the duplications, edited the writeup based on the comments (will open another PR for the motivation section). Please let me know if further updates are required. Thanks!

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -1,37 +1,59 @@
 
 # Table of Contents
 
+- [Table of Contents](#table-of-contents)
```
```suggestion
```

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -1,37 +1,59 @@
 # Abstract
 
-`WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes.
+`WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes. It also supports pagination for more efficient querying of historical messages where a querying node can set the number of `WakuMessage`s to return and a cursor to start from. The queried node MUST then use this information to return messages in the queried range. The query response ...
```
I'd either move or remove:

> The queried node MUST then use this information to return messages in the queried range. The query response also embodies an updated cursor which enables the querying node resume its subsequent requests in an iterative manner.

It is a bit too detailed/spec-like for an abstract/executive summary with language such as MUST etc.

As a reader skimming an abstract, I only want to know roughly what functionality is provided and why it exists, not know how to use it or exact guarantees for API calls. The shorter the better!

staheri14 replied:
Sure, addressed

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +63,37 @@ message HistoryRPC {
 
-##### HistoryRPC
+### Index
+To perform pagination, each `WakuMessage` stored at a node running the store protocol is associated with a unique `Index` that encapsulates the following parts. 
```
```suggestion
### Index

To perform pagination, each `WakuMessage` stored at a node running the store protocol is associated with a unique `Index` that encapsulates the following parts. 
```

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +63,37 @@ message HistoryRPC {
-A node MUST send all History messages (`HistoryQuery`, `HistoryResponse`) wrapped inside a
-`HistoryRPC`.
+### PagingInfo
+`PagingInfo` holds the information required for pagination. It consists of the following components.
```
```suggestion
### PagingInfo

`PagingInfo` holds the information required for pagination. It consists of the following components.
```

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +63,37 @@ message HistoryRPC {
-2.0.0-beta1
-Initial draft version. Released [2020-10-06](https://github.com/vacp2p/specs/commit/75b4c39e7945eb71ad3f9a0a62b99cff5dac42cf)
+2.0.0-beta2
+Initial draft version. Released [2020-10-21](https://github.com/vacp2p/specs/commit/75b4c39e7945eb71ad3f9a0a62b99cff5dac42cf)
```
The way we do releases is:

1) People approve PR
2) Squash commits in PR
3) Add one more commit to PR with commit info to previously squashed PR
4) Merge rebase

This means the PR results in two commits on master, with the commit hash of a release leading to all the changes done in that release (i.e. this PR).

---

Also the 2.0.0-beta1 should still exist, just appended to in reverse chronological order. Then the beta2 will have a changelog with something like:

- Added pagination support
- Clarified X

staheri14 replied:
Thanks for the clarification :)

oskarth commented on specs/waku/v2/waku-v2.md:
```diff
@@ -1,6 +1,6 @@
 ---
 title: Waku
-version: 2.0.0-beta1
+version: 2.0.0-beta2
```
Not sure we need to bump this in PR, though it might be a good idea to update the reference to store protocol here to beta2

Honestly the versioning of main spec isn't obvious right now, so we might need to figure out what the best flow looks like here. It doesn't immediately correspond to a protocol string, and other specs will move out of sync, etc. For now I'd leave it be and we can bump this separately.

oskarth commented on specs/waku/v2/waku-v2.md:
```diff
@@ -122,7 +122,7 @@ See [bridge spec](waku-bridge.md) for details on a bridge mode.
 ## Changelog
 
-### 2.0.0-beta1
+### 2.0.0-beta2
```
Same here, also see above re changelog. We don't want to overwrite it, the goal is to have a short history to see what changed etc between versions.

See https://specs.vac.dev/specs/waku/waku.html#changelog for an example

oskarth commented on wordlist.txt:
```diff
@@ -135,6 +138,7 @@ rlp
 rlpx
 RLPx
 rpc
+sanaz
```
Huh, surprised it is lowercase, does that work? I thought the spell checker was case sensitive

staheri14 replied:
Apparently yes!

oskarth reviewed (commented):
Thanks! Almost there, just a few minor details outstanding

decanus commented on specs/waku/v2/waku-store.md:
```diff
@@ -1,37 +1,59 @@
 # Wire Specification
-
-Peers communicate with each other using a request / response API. The messages sent are Protobuf RPC messages.
+Peers communicate with each other using a request/response API. The messages sent are Protobuf RPC messages. The followings are the specifications of the Protobuf messages. 
```
```suggestion
Peers communicate with each other using a request / response API. The messages sent are Protobuf RPC messages. The followings are the specifications of the Protobuf messages. 
```

Not that important but I would readd the spaces.

decanus commented on specs/waku/v2/waku-store.md:
```diff
@@ -1,37 +1,59 @@
+
+message PagingInfo {
+  int64 pageSize=1 
+  Index cursor=2
```
```suggestion
  bytes digest = 1;
  float receivedTime = 2;
}

message PagingInfo {
  int64 pageSize = 1;
  Index cursor = 2;
```

decanus commented on specs/waku/v2/waku-store.md:
```diff
@@ -1,37 +1,59 @@
   repeated string topics = 2;
+  optional PagingInfo pagingInfo = 3 // used for pagination
+
 }
```
```suggestion
  optional PagingInfo pagingInfo = 3; // used for pagination
}
```

decanus commented on specs/waku/v2/waku-store.md:
```diff
@@ -1,37 +1,59 @@
 
 message HistoryResponse {
   repeated WakuMessage messages = 2;
+  optional PagingInfo pagingInfo = 3 // used for pagination
```
```suggestion
  optional PagingInfo pagingInfo = 3; // used for pagination
```

decanus commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +63,37 @@ message HistoryRPC {
 
-The `topics` field MUST indicate the list of topics to query.
+- The `topics` field MUST indicate the list of topics to query.
+- `PagingInfo` holds the information required for pagination.  Its `pageSize` field indicates the number of  `WakuMessage`s to be included in the corresponding `HistoryResponse`. If the `pageSize` is omitted or exceeds an upper limit then a “default page size” shall be used instead. In the forward pagination request, the `messages` field of the `HistoryResponse` shall contain `pageSize` many waku ...
```
```suggestion
- `PagingInfo` holds the information required for pagination.  Its `pageSize` field indicates the number of  `WakuMessage`s to be included in the corresponding `HistoryResponse`. If the `pageSize` is omitted or exceeds an upper limit then a "default page size" shall be used instead. In the forward pagination request, the `messages` field of the `HistoryResponse` shall contain `pageSize` many waku messages whose `Index` are larger than the given `cursor` (and vise versa for the backward pagination). Note that the `cursor` of a `HistoryQuery` may be empty (e.g., for the initial query), as such, the store node, depending on whether the  `direction` is `BACKWARD` or `FORWARD` returns the last or the first `pageSize` waku messages, respectively.
```

oskarth reviewed (approved):
LGTM! Please rebase and squash, then add another commit with release hash

staheri14 commented:
@decanus would you please review the changes, thanks!

kdeme commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +63,45 @@ message HistoryRPC {
 
-The `topics` field MUST indicate the list of topics to query.
+- The `topics` field MUST indicate the list of topics to query.
+- `PagingInfo` holds the information required for pagination.  Its `pageSize` field indicates the number of  `WakuMessage`s to be included in the corresponding `HistoryResponse`. If the `pageSize` is omitted or exceeds an upper limit then a "default page size" shall be used instead. In the forward pagination request, the `messages` field of the `HistoryResponse` shall contain `pageSize` many waku ...
```
This reads a bit difficult and is probably grammatically not fully correct.
Perhaps this is better? Not sure.
```suggestion
- `PagingInfo` holds the information required for pagination.  Its `pageSize` field indicates the number of  `WakuMessage`s to be included in the corresponding `HistoryResponse`. If the `pageSize` is omitted or exceeds an upper limit then a "default page size" shall be used instead. In the forward pagination request, the `messages` field of the `HistoryResponse` shall contain at maximum the `pageSize` amount of waku messages whose `Index` values are larger than the given `cursor` (and vise versa for the backward pagination). Note that the `cursor` of a `HistoryQuery` may be empty (e.g., for the initial query), as such, the store node, depending on whether the  `direction` is `BACKWARD` or `FORWARD` returns the last or the first `pageSize` waku messages, respectively.
```

kdeme reviewed (approved):
Sorry for the late review, lgtm.

I do have a general question though:
It is written implicitly that `Index` is used to order the messages because it is stated that a request should provide messages with a larger or smaller index depending the direction.
Now I assume that the ordering happens by checking the `receivedTime` + `digest` combination where the `receivedTime` is the "more significant part" so that you can practically order them by time if you want.

Now it is nowhere stated that it should be done like this. Is this something we deliberately want to leave open? To me it seems useful to add this when you also define out of which fields the `Index` consists.

staheri14 commented:
> Sorry for the late review, lgtm.
> 
> I do have a general question though:
> It is written implicitly that `Index` is used to order the messages because it is stated that a request should provide messages with a larger or smaller index depending the direction.
> Now I assume that the ordering happens by checking the `receivedTime` + `digest` combination where the `receivedTime` is the "more significant part" so that you can practically order them by time if you want.
> 
> Now it is nowhere stated that it should be done like this. Is this something we deliberately want to leave open? To me it seems useful to add this when you also define out of which fields the `Index` consists.

@kdeme I am totally fine with adding more details about the sorting logic, what do you think @oskarth?

oskarth commented:
We can add it as a MAY or RECOMMENDED to leave open for other mechanisms (eg sender determined, see other issues for details)

staheri14 commented:
> We can add it as a MAY or RECOMMENDED to leave open for other mechanisms (eg sender determined, see other issues for details)

@kdeme following my meeting with  @oskarth , I am going to add the sorting details inside the codebase of pagination, as the description of the pagination proc (i.e., https://github.com/status-im/nim-waku/blob/af79c14e41c3eab356aa3d9ee02de2ba70a1f343/waku/protocol/v2/waku_store.nim#L184), let me know whether this addresses your comment?

oskarth commented on specs/waku/v2/waku-store.md:
```diff
@@ -41,30 +63,45 @@ message HistoryRPC {
-Initial draft version. Released [2020-10-06](https://github.com/vacp2p/specs/commit/75b4c39e7945eb71ad3f9a0a62b99cff5dac42cf)
+
+### 2.0.0-beta2 
+Released [2020-10-27]()
```
Squash to one commit and add release commit, then we can merge!

kdeme commented:
> > We can add it as a MAY or RECOMMENDED to leave open for other mechanisms (eg sender determined, see other issues for details)
> 
> @kdeme following my meeting with @oskarth , I am going to add the sorting details inside the codebase of pagination, as the description of the pagination proc (i.e., https://github.com/status-im/nim-waku/blob/af79c14e41c3eab356aa3d9ee02de2ba70a1f343/waku/protocol/v2/waku_store.nim#L184), let me know whether this addresses your comment?

@staheri14 It would be nice to write down the reasoning to "why?" then. For me it still sounds odd to not have this defined. As it means a client can connect to some other implementation and have no idea of the ordering. E.g. maybe it expects to receive the newest messages first, but it doesn't.

staheri14 commented:
> @staheri14 It would be nice to write down the reasoning to "why?" then. For me it still sounds odd to not have this defined. As it means a client can connect to some other implementation and have no idea of the ordering. E.g. maybe it expects to receive the newest messages first, but it doesn't.

@kdeme Sorry, I forgot to explain "why", indeed, the sorting algorithm is an implementation decision and it is better to not to enforce a particular approach at the specs level, also, in order to keep the specs less detailed, the decision was to describe the sorting in the code rather than the specs. However, I agree that a quick overview of the sorting  would be helpful and meaningful. I added an overview of the sorting in my recent commits . Please let me know of further details are needed. Upon your approval, I will merge the PR.

oskarth commented:
@staheri14 Looks like we all good to squash, add commit info and merge

staheri14 committed:
adds pagination specs

adds some edits


fixes minor typo


WIP: integrates the pagination into store protocol


updates wordlist.txt


WIP: minor updates


updates wordlist


minor updates


updates TOC


updates the specs


fixes a typo


beta1 to beta2


deletes waku-store-pagination.md 


updates TOC with an autogenerated markdown-toc 


Update specs/waku/v2/waku-store.md

Co-authored-by: Oskar Thorén <ot@oskarthoren.com>
Update specs/waku/v2/waku-store.md

Co-authored-by: Oskar Thorén <ot@oskarthoren.com>
Update specs/waku/v2/waku-store.md

Co-authored-by: Oskar Thorén <ot@oskarthoren.com>
Update specs/waku/v2/waku-store.md

Co-authored-by: Oskar Thorén <ot@oskarthoren.com>
Update specs/waku/v2/waku-store.md

Co-authored-by: Oskar Thorén <ot@oskarthoren.com>
itemizes the message fields of HistoryRPC 


some updates


Update specs/waku/v2/waku-store.md

Co-authored-by: Dean Eigenmann <7621705+decanus@users.noreply.github.com>
WIP: initial reorganization 


WIP: shortens the pagination spec


removes some comments and adds more details to the wire spec


minor updates


beta1 to beta2


converts direction to enum


edits the Index description 


Update specs/waku/v2/waku-store.md

Co-authored-by: Oskar Thorén <ot@oskarthoren.com>
shortens the abstract


Update specs/waku/v2/waku-store.md

Co-authored-by: Oskar Thorén <ot@oskarthoren.com>
Update specs/waku/v2/waku-store.md

Co-authored-by: Oskar Thorén <ot@oskarthoren.com>
edits changelog


Update specs/waku/v2/waku-store.md

Co-authored-by: Dean Eigenmann <7621705+decanus@users.noreply.github.com>
Update specs/waku/v2/waku-store.md

Co-authored-by: Dean Eigenmann <7621705+decanus@users.noreply.github.com>
Update specs/waku/v2/waku-store.md

Co-authored-by: Dean Eigenmann <7621705+decanus@users.noreply.github.com>
Update specs/waku/v2/waku-store.md

Co-authored-by: Dean Eigenmann <7621705+decanus@users.noreply.github.com>
minor


minor


reformats the change log


Update specs/waku/v2/waku-store.md

Co-authored-by: Dean Eigenmann <7621705+decanus@users.noreply.github.com>
edits version numbers


updates the release date


removes Table of content from Table of content!
Update specs/waku/v2/waku-store.md

Co-authored-by: Kim De Mey <kim.demey@gmail.com>
edits the wordlist


explains the case where pageSize is zero


Update waku-store.md

Update waku-v2.md
adds an overview of the sorting approach


minor


Update waku-store.md

staheri14 committed:
adds commit info
